### ISSUE TITLE
Function findLast crashes when defining 'startIndex'

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#831]
**Reported by tobolar on 20 Sep 2012 07:47 UTC**
Function Modelica.Utilities.Strings.findLast crashes when using parameter 'startIndex' and 
startIndex > length(string)-length(searchString)+1

Compare call

`Modelica.Utilities.Strings.findLast("string to be analysed", "to be", Modelica.Utilities.Strings.length("string to be analysed")-Modelica.Utilities.Strings.length("to be")+1, false);`

with (startIndex is of 1 higher)

`Modelica.Utilities.Strings.findLast("string to be analysed", "to be", Modelica.Utilities.Strings.length("string to be analysed")-Modelica.Utilities.Strings.length("to be")+2, false);`

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/831

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 20 Sep 2012 10:06 UTC**
I agree, the condition basically means, that the start string will never be found.

The function should be modified. I think the best case would be to not use startIndex == 0 for the default argument. The documentation already states that startIndex defaults to length(string) (which it does by switching the 0 case).

This is my suggestion:

```mo
  function findLast "Find last occurrence of a string within another string"
    extends Modelica.Icons.Function;
    input String string "String that is analyzed";
    input String searchString "String that is searched for in string";
    input Integer startIndex(min=0)=length(string)
      "Start search at index startIndex. Defaults to length(string)";
    input Boolean caseSensitive=true
      "= false, if lower and upper case are ignored for the search";
    output Integer index
      "Index of the beginning of the last occurrence of 'searchString' within 'string', or zero if not present";
  protected
    Integer lenString = length(string);
    Integer lenSearchString = length(searchString);
    Integer i;
  algorithm
    i := min(lenString-lenSearchString+1, startIndex);
    index := 0;
    while i >= 1 loop
       if isEqual(substring(string,i,i+lenSearchString-1),
                  searchString, caseSensitive) then
          index := i;
          i := 0;
       else
          i := i-1;
       end if;
    end while;
    annotation (
  Documentation(info="<html>
<h4>Syntax</h4>
<blockquote><pre>
index = Strings.<b>findLast</b>(string, searchString);
index = Strings.<b>findLast</b>(string, searchString,
                         startIndex=length(string), caseSensitive=true,
</pre></blockquote>
<h4>Description</h4>
<p>
Finds first occurence of \"searchString\" within \"string\"
when searching from the last character of \"string\"
backwards, and return the corresponding index.
Start search at index \"startIndex\" (default = length(string)).
If the optional argument \"caseSensitive\" is false, lower
and upper case are ignored for the search.
If \"searchString\" is not found, a value of \"0\" is returned.
</p>
</html>"));
  end findLast;
```

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 20 Dec 2012 13:47 UTC**
Housekeeping, defaulting all new tickets to [milesstone:\*undecided\*] (is normally done automatically)

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 30 Mar 2013 12:53 UTC**
Fixed in a9ab374a3e5f570ebb8e166b7e20e63f0c36f374:
The first index has been corrected, so that the reported situation can no longer occur.

The solution proposed in the ticket is not so good, because MSL 3.2 implementation has already the option that startIndex=0 means to start search at length(string), so this must be still supported

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 30 Mar 2013 12:54 UTC**
